src/pybind/mgr/dashboard/services/oauth2.py

```diff
@@ -0,0 +1,132 @@
+import json
+import logging
+from typing import Dict, List
+from urllib.parse import quote
+
+import cherrypy
+import requests
+from cherrypy import _ThreadLocalProxy
+
+from dashboard.controllers.auth_classes import decode_jwt_segment
+
+from .. import mgr
+from ..services.access_control import Role, User, UserAlreadyExists
+from ..tools import prepare_url_prefix
+
+logger = logging.getLogger('OAuth2')
+
+
+class OAuth2:
+    @classmethod
+    def get_token(cls, request: _ThreadLocalProxy) -> str:
```

Comment:
There's a lot of code in this class and I have the feeling that some amount of it could be shared with local/saml2 implementations? If that's the case, I'd suggest that we have a base class with common methods.

Reply:
There are some similarities between this and the local/saml2 as you mentioned.
In the `services/oauth2.py` I can see the following methods that could be inherted for all protocols: `get_token` `set_token`, `get_token_payload`, `set_token_payload`. But even this "common" methods differ a bit from the current local implementation. A part from these, the rest are particular to the oauth2 protocol.

For sure we could have a base class and encapsulate some of the code there and expand on each of the protocols but I think that would required a lot of changes on the current local and saml2 implementation, and I would like to avoid that, for now at least.

Reply:
Ok, that's the benefit of defining an abstract interface: you move the common methods to the ABC class, and then force the implementation in the derived classes. If there's some code that could also be used, you can always inherit the method as well:
```py
class SSOAuth(BaseAuth):
   def get_token(self):
     # logic to parse token from header/cookie
     return token

class OAuth2(SSOAuth):
  def get_token(self):
    token = super().get_token()
    # do extra stuff with the token
```

Reply:
[This](https://github.com/ceph/ceph/pull/58456#discussion_r1756205897) follows on with the same thing.

Some refactoring on the JwtManager is neeeded to achive this

Reply:
Ok!